### ISSUE TITLE
Consider support for Bridgy publish to Pixelfed (Fediverse) instances

### ISSUE THREAD

**Issue by rretsiem** [snarfed/bridgy#927]
To publish images to [Pixelfed](https://pixelfed.org) (Fediverse), support for those instances would be a great addition. Since the [Pixelfed API](https://docs.pixelfed.org/technical-documentation/api-v1.html#authorization) is based on Mastodon V1 API, I hope that therefore not a complete new development is necessary. At least it sounds like a “low hanging fruit” 😉

A possible use case would be to publish photos posts to Pixelfed instances and also enabling Webmentions to retrieve reactions to those posts.

If this is already possible with the “Mastodon” function, a hint on how to use it would be helpful.

(Originally published at: https://renem.net/notes/github/2020-03-16-1/)

**Comment by snarfed** [comment]
yes! great idea. thanks for filing!

(i assume you've looked at https://fed.brid.gy/ ? it's another good approach, but not for the faint of heart.)

when i first implemented mastodon, i tried pixelfed, and it wasn't quite compatible enough, so i detect and block it explicitly: https://github.com/snarfed/oauth-dropins/commit/8b8405aa0c5ae14d11eade3f16b6c5535e51fd52 . if it's more compatible now, it might be as easy as removing that block and trying again!

any interest in trying? i'm happy to help you get the dev environment set up!

**Comment by rretsiem** [comment]
Thanks for the info. I try to get GAE up and running on my local MacBook and see what I can achieve. But I'm honest, I'm just at the beginning for now 😀
For example, I still don't know why your answer to my Issue isn't sent back to my webmention endpoint. Brid.gy says "no webmention targets". (Not a topic of the current issue, so maybe in IRC/Slack).

**Comment by snarfed** [comment]
huh. thanks for reporting, and sorry for the trouble! that may indeed be a bug. i'll file a new issue for it.

**Comment by snarfed** [comment]
note to myself: my pixelfed account, mostly just for testing this, is https://piconic.co/snarfed .

**Comment by snarfed** [comment]
current status: debugging pixelfed's oauth. i've gotten through to requesting the access token, but pixelfed 400s that request with _The authorization grant type is not supported by the authorization server._ (i'm passing `grant_type=authorization_code`.) not sure which other grant type i'd use. i'll ask them.

**Comment by snarfed** [comment]
filed an issue with more details: https://github.com/pixelfed/pixelfed/issues/2102

**Comment by hirusi** [comment]
Looks like much of the work is done in https://github.com/snarfed/bridgy/commit/bd900cf6436e221eb36e6179b163de13720fcd07 - is this considered ready for production yet?

**Comment by snarfed** [comment]
you're right! the bridgy work is largely done, in that commit. deploying to production is currently blocked on a [handful of pixelfed bugs that affect both backfeed and publish](https://github.com/pixelfed/pixelfed/issues?q=is%3Aissue+author%3Asnarfed). fortunately @dansup et al have usually been quick to respond to these, so i'm optimistic those will get fixed pretty soon.

**Comment by hirusi** [comment]
Not a problem and thanks for the status update. Really do appreciate all that you do!

**Comment by snarfed** [comment]
[interesting news](https://chat.indieweb.org/dev/2020-05-28/1590648527037900):
> [dansup] After some thought, I've decided to add Webmention support to Pixelfed, users will be able to opt out)
> ...
> [snarfed] dansup++ webmention support in pixelfed! exciting! i assume you just mean directly, ie receiving and sending with links to/from posts? not bridgy-style backfeed for posts that are syndicated from people's web sites?

**Comment by bekopharm** [comment]
Pixelfed evolved over the last year. Maybe it's time to revise this?

**Comment by snarfed** [comment]
Sure! I haven't followed its evolution. How do you think we should revise this?

**Comment by bekopharm** [comment]
> blocked on Pixelfed bugs

I wonder if the bugs do still exist.

**Comment by snarfed** [comment]
[Looks like at least half of them do.](https://github.com/pixelfed/pixelfed/issues?q=is%3Aissue+author%3Asnarfed)

**Comment by snarfed** [comment]
Hey @dansup any chance you could look at the bugs ^ blocking Bridgy Pixelfed support? I'd love to get this moving again!

...or if you're close to shipping built in webmention support, that'd be awesome, and I could close this altogether!

**Comment by kristofzerbe** [comment]
Is there a chance to revisit the topic?

**Comment by kristofzerbe** [comment]
I have questioned @pixelfed@mastodon.social ... Lets see

**Comment by StarrWulfe** [comment]
@snarfed.org@snarfed.org [@dansup](https://mastodon.social/@dansup)

…any updates on this?🙏🏾

I just found out this wasn’t working myself.

_In reply to[https://github.com/snarfed/bridgy/issues/927#issuecomment-911780729](https://github.com/snarfed/bridgy/issues/927#issuecomment-911780729)._

(Originally published at: https://starrwulfe.xyz/b/jf8)

**Comment by snarfed** [comment]
Good question! Some of the bugs ^ are still open, but maybe they've improved? Feel free to try and let us know how it goes!

**Comment by StarrWulfe** [comment]
![image](https://github.com/snarfed/bridgy/assets/445009/80189686-e6f6-40cc-9947-98784645a98e)
@snarfed
Bridgy still comes back with errors as if Pixelfed isn’t an ActivityPub server. I’m not sure what behavior is being expected, but Pixelfed (at least the flagship server at pixelfed.social) isn’t doing what Bridgy expects…

**Comment by snarfed** [comment]
Looked into this a bit. [Sounds like Pixelfed fixed some of these bugs!](https://mastodon.cloud/@dansup@mastodon.social/111513464667133152) Not all though, evidently. Looks like we're still stuck on OAuth scope parity with Mastodon though:
* Does Pixelfed not support [fine-grained scopes](https://docs.joinmastodon.org/api/oauth-scopes/#granular)? When I try to use eg `scope=read:accounts` with https://piconic.co/oauth/authorize, it redirects back with HTTP 400 `invalid_scope The requested scope is invalid, unknown, or malformed`
* https://github.com/pixelfed/pixelfed/issues/2106 is evidently still outstanding, we need that to work too

**Comment by kristofzerbe** [comment]
Please be so kind and take a look at this. I would love to see Pixelfed bridging...

**Comment by snarfed** [comment]
@kristofzerbe I'd love it too! Sadly we're still blocked on a handful of Pixelfed bugs and API problems, eg the two in https://github.com/snarfed/bridgy/issues/927#issuecomment-1837292601.

**Comment by kristofzerbe** [comment]
@snarfed I don't give up asking ... ;)

**Comment by snarfed** [comment]
Progress! @dansup fixed https://github.com/pixelfed/pixelfed/issues/2106, which was our one main remaining blocker here.

Now the work is on our side. I'd need to add Pixelfed as a new silo, since it still doesn't support the same fine-grained OAuth scopes as Mastodon. That'll take some work, and I don't know when I'd prioritize it, but it's all straightforward, and the oauth-dropins and granary Pixelfed classes already exist. https://bridgy.readthedocs.io/#adding-a-new-silo describes the remaining work if anyone is interested in trying!
